packages/salesforcedx-vscode-apex/src/languageUtils/languageClientUtils.ts

```diff
@@ -104,3 +119,137 @@ export const getExceptionBreakpointInfo = async (): Promise<{}> => {
   }
   return Promise.resolve(response);
 };
+
+export const restartLanguageServerAndClient = async (extensionContext: vscode.ExtensionContext) => {
+  const removeIndexFiles = await vscode.window.showInformationMessage(
+    nls.localize('apex_language_server_restart_dialog_prompt'),
+    nls.localize('apex_language_server_restart_dialog_clean_and_restart'),
+    nls.localize('apex_language_server_restart_dialog_restart_only')
+  );
+  //get the client and status bar item
+  const alc = LanguageClientUtils.getInstance().getClientInstance();
+  let statusBarInstance = LanguageClientUtils.getInstance().getStatusBarInstance();
+  if (statusBarInstance === undefined) {
+    statusBarInstance = new ApexLSPStatusBarItem();
+    LanguageClientUtils.getInstance().setStatusBarInstance(statusBarInstance);
+  }
+  if (alc !== undefined) {
+    statusBarInstance.restarting(); // update the status to user that we're restarting
+    // Stop the client and server with the safe stop() routine from the client.
+    // Server is stopped automatically as part of that procedure
+    try {
+      //if we try and restart during another restart or with a startFailed status, we'll throw an exception
+      await alc.stop();
+    } catch (error) {
+      //if the client is in the state of startFailed, we need to kill the LSP jar and then move on.
+      //todo: do we want to kill LSP jar manually here?
+      vscode.window.showWarningMessage(
+        nls.localize('apex_language_server_restart_dialog_restart_only') + error.message
+      );
+    }
+    if (removeIndexFiles === nls.localize('apex_language_server_restart_dialog_clean_and_restart')) {
+      //delete the apex DB file if we need a clean slate
+      await removeApexDB();
+    }
+    //Need a half second timeout for the creation of a new language client so that we give the previous LS a moment to complete shutdown
+    setTimeout(createLanguageClient, 500, extensionContext, statusBarInstance);
+  }
+};
+
+const removeApexDB = async () => {
+  if (vscode.workspace.workspaceFolders) {
+    const wsrf = vscode.workspace.workspaceFolders[0].uri;
+    const toolsUri = vscode.Uri.joinPath(wsrf, '.sfdx', 'tools');
+    try {
+      const entries = await vscode.workspace.fs.readDirectory(toolsUri);
+      const releaseFolders = entries
+        .filter(([name, type]) => type === vscode.FileType.Directory && /^\d{3}$/.test(name))
+        .map(([name]) => name);
+      for (const folder of releaseFolders) {
+        const folderUri = vscode.Uri.joinPath(toolsUri, folder);
+        await vscode.workspace.fs.delete(folderUri, { recursive: true, useTrash: true });
```

Comment:
I'm used to having the `no-await-in-loop` rule in place so that you parallelize everything by default.  

I'm guessing these are the 252, 254 etc folders so there's not too many of them that it would matter.

Reply:
Yeah I think unless there is something that's gone horribly wrong, users would have at most 2 folders (an old release and the new release) where we could blow both away and start fresh with a regenerated folder from the LSP. This shouldn't be handling more than couple at any given time.